### PR TITLE
584 varun  abstract clustering ml

### PR DESCRIPTION
**Pull request by varun646**
# Description:
- Add interface for clustering models
- Add BERTopicModel class that adheres to interface
- Add LDAModel class that adheres to interface
# Demonstration:

Screenshot 2024-11-21 at 12 57 13 PM

### CODE REVIEW

File: ClusteringModel.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+from abc import ABC, abstractmethod
+
+
+class ClusteringModel(ABC):
+    @abstractmethod
+    def getClusters(self, documents):
```
Comment (davidbrownell) [inline ClusteringModel.py]: Consider adding type hints (especially for interface definitions) as it makes it much easier to add new derived classes and also ensure that existing derived classes are correct.

File: BERTopicModel.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+                         calculate_probabilities=calculate_probabilities,
+                         seed_topic_list=seed_topic_list,
+                         zeroshot_topic_list=zeroshot_topic_list,
+                         zeroshot_min_similarity=zeroshot_min_similarity,
+                         embedding_model=embedding_model, umap_model=umap_model,
+                         hdbscan_model=hdbscan_model,
+                         vectorizer_model=vectorizer_model, ctfidf_model=ctfidf_model,
+                         representation_model=representation_model, verbose=verbose)
+
+    def getClusters(self, documents):
```
Comment (davidbrownell) [inline BERTopicModel.py:40]: Consider using `override` from the `typing` module to indicate that the method is derived.

File: LDAModel.py [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+import numpy as np
+from gensim.models import ldamodel
+import gensim.corpora as corpora
+import octis.configuration.citations as citations
+import octis.configuration.defaults as defaults
+from gensim.models import TfidfModel
+
+
+class LDAModel(ClusteringModel, AbstractModel):
+    id2word = None
```
Comment (davidbrownell) [inline LDAModel.py:12]: These are class instance variables but seem to be used as instance variables. If they really are class instance variables, remove the code that sets the values within instance methods.

**Review by davidbrownell — CHANGES_REQUESTED** [review]

File: ClusteringModel.py (hunk 1 above)
Comment (varun646) [reply ClusteringModel.py]: makes sense - will add

**Review by davidbrownell — CHANGES_REQUESTED** [review]

File: src/ClusteringModelInterface.py [hunk 4]
```diff
@@ -1,11 +1,20 @@
 from abc import ABC, abstractmethod
+from os import PathLike
 
 
 class ClusteringModelInterface(ABC):
     @abstractmethod
-    def getClusters(self, documents):
+    def getClusters(self, datapath: PathLike):
```
[the remark this answers is not in the document]
Comment (davidbrownell) [reply src/ClusteringModelInterface.py]: Gotcha. I see that now. Type hints in the derived classes would be helpful as well :) (but that can be done later)

**Review by davidbrownell — APPROVED** [review]

**Pull request merged**